### PR TITLE
Electrostatics

### PR DESCRIPTION
**Pull request by dylanbeaudette**
New label overlap adjustments, with predictable output. Numerous small bug fixes in `plotSPC()`. Quickly create tiny SPCS from text or list templates with `quickSPC().

Addressing tasks from #285.

### CODE REVIEW

**Comment by dylanbeaudette** [comment]
Hey @brownag take a peak at this if you have time.

**Comment by brownag** [comment]
> Hey @brownag take a peak at this if you have time.

I pulled this branch and checked a few things, so far no breakages or unexpected results. 
There are a couple items I want to look at and check but can complete review within a day or so.

If I am not mistake, with this PR changes compared against master I saw an order of magnitude _less_ RAM usage  when plotting with a large (way too large to be viewable) SPC.

**Comment by dylanbeaudette** [comment]
> I pulled this branch and checked a few things, so far no breakages or unexpected results. There are a couple items I want to look at and check but can complete review within a day or so.

Thanks. I recognize that A LOT got lumped into this single PR--that was originally focused on label placement. No rush. This and some NCSP-related documentation are the last (ha) things stopping a 2.0 release.
 
> If I am not mistake, with this PR changes compared against master I saw an order of magnitude _less_ RAM usage when plotting with a large (way too large to be viewable) SPC.

Thanks for checking that; you will have to show me how you did that sometime. I'm not surprised though. The new default method replaces non-deterministic perturbations (with constraints) with simpler, physical model. If I have time, I'll put together a short vignette.

**Comment by brownag** [comment]
I think I have completed my initial review. I have not seen any glaring problems.

I am wondering if the following is expected and/or whether it can be optimized to work better.
See Tujunga ^A horizon bottom depth label. 

image

Should the 4 be repelled a bit more from the 0 with default settings? Is this affected by the representation of the boundary topography / are those depths considered? This is from the last example in https://github.com/ncss-tech/sharpshootR/blob/master/misc/soilTaxonomyDendrogram-examples.R

---

The observed RAM differences were just apparent via `bench::mark()` I was only looking for major changes in resource consumption/processing time to make sure that intractable cases didn't process for too long.

With my current selected set (n=1164 pedons) 5.24s and 151MB memory consumed:
```r
f <- fetchNASIS()
bench::mark(plot(f))
#> # A tibble: 1 × 13
#>  expre…¹   min median itr/s…² mem_a…³ gc/se…⁴ n_itr  n_gc total…⁵ result       memory     time      
#>  <bch:e> <bch> <bch:>   <dbl> <bch:b>   <dbl> <int> <dbl> <bch:t> <list>       <list>     <list>    
#>1 plot(f) 5.24s  5.24s   0.191   151MB    2.29     1    12   5.24s <named list> <Rprofmem> <bench_tm>
```

Whereas with current master branch w/ the simulated annealing (rather than electrostatic method):

```r
#> # A tibble: 1 × 13
#>   expre…¹   min median itr/s…² mem_a…³ gc/se…⁴ n_itr  n_gc total…⁵ result       memory     time      
#>   <bch:e> <bch> <bch:>   <dbl> <bch:b>   <dbl> <int> <dbl> <bch:t> <list>       <list>     <list>    
#> 1 plot(f) 5.53s  5.53s   0.181  2.16GB    6.50     1    36   5.53s <named list> <Rprofmem> <bench_tm>
```

So, appears to be a very significant improvement in that degenerate (large SPC) case!

**Review by brownag — APPROVED** [review]

File: R/aqp-label-placement-solvers.R [hunk 1]
```diff
@@ -17,62 +115,124 @@
...
 }
 
 
 
-## TODO: allow for min.x | min.y as defined in parent function
-
 
 #' @title Label placement based on a simulation of electrostatic forces
+#' 
+#' @description This function will attempt
```
[the remark this answers is not in the document]
Comment (dylanbeaudette) [reply R/aqp-label-placement-solvers.R:144]: Thanks, fixed. I plan to introduce more detailed documentation via vignette.

File: R/aqp-label-placement-solvers.R [hunk 2]
```diff
@@ -1,12 +1,110 @@
...
+#' 
+#' @export
+#'
+#' @examples 
+#' 
+#' x <- c(1, 2, 3, 3.4, 3.5, 5, 6, 10)
+#' 
+#' findOverlap(x, thresh = 0.5)
+#'
+findOverlap <- function(x, thresh) {
```
[the remark this answers is not in the document]
Comment (dylanbeaudette) [reply R/aqp-label-placement-solvers.R:21]: It is, will mark as such.

**Comment by dylanbeaudette** [comment]
Thanks for the detailed feedback!

As for the TUJUNGA example: that is the expected output because `plotSPC()` uses `fixLabelCollisions = FALSE` by default. We might consider changing that in this PR, but I hadn't yet considered the implications. Previously, I'd expected only setting `fixLabelCollisions = TRUE` when necessary, via argument. I'm open to suggestions. When setting it to `TRUE`, the results look OK.

image

For the record, the horizon depths are always positioned at the horizon depths as-specified in the SPC and not adjusted according to boundary information.

**Comment by brownag** [comment]
> As for the TUJUNGA example: that is the expected output because `plotSPC()` uses `fixLabelCollisions = FALSE` by default. We might consider changing that in this PR, but I hadn't yet considered the implications. Previously, I'd expected only setting `fixLabelCollisions = TRUE` when necessary, via argument. I'm open to suggestions. When setting it to `TRUE`, the results look OK.

Ah, my mistake, I misinterpreted and jumped in to test it because I had seen notes about it being broken by changes in this PR. I should have realize; the result makes sense then! 

I suppose needing to use an argument is fine but I am personally in favor of the defaults being as close to ready-to-use as possible. It doesn't seem like there would be a performance reason not to use it whenever you could, or at least have `fixLabelCollisions = hz.depths` in `plotSPC()`? When `hz.depths=TRUE` it makes the label overlap problem particularly difficult, in other cases maybe it is less important.


> image
> 
> For the record, the horizon depths are always positioned at the horizon depths as-specified in the SPC and not adjusted according to boundary information.

Yeah, from the figure its clear they are based at the bottom depth/midpoint. I wouldn't expect this method to use boundary information per se, but was curious whether any thought to non-default boundary geometry was considered. It does look fine with the option turned on. All good there, thanks!

**Comment by dylanbeaudette** [comment]
Good idea on setting `fixLabelCollisions = hz.depths`, I like that--especially with the more efficient method. I was originally hesitant due to the non-deterministic output.

A remaining TODO, after this PR is a simple heuristic for selecting a `q` value. It depends on a couple of factors. The following will be the start of a short vignette.

image

**Pull request merged**